nemo/collections/nlp/data/dialogue/data_processor/custom_data_processor.py

```diff
@@ -0,0 +1,210 @@
+from nemo.collections.nlp.data.dialogue.input_example.input_example import DialogueInputExample
+from nemo.utils import logging
+
+
+class DialogueCustomDataProcessor(DialogueDataProcessor):
+    """ Data processor for a custom dialog bot dataset """
+
+    def __init__(self, data_dir: str, tokenizer: object = None, cfg=None):
+        """ Constructs the dialog data processor """
+
+        self.data_dir = data_dir
+        self.cfg = cfg
+        self._task_name = self.cfg.task_name
+        self._tokenizer = tokenizer
+
+    def get_dialog_examples(self, dataset_split) -> List[object]:
+        """
+        Reads dialogue examples from self.data_dir/dataset_split/examples.json
+        and formats into SGD structure
+        """
+
+        dialogues_filepath = os.path.join(self.data_dir, dataset_split, "examples.json")
+        if not os.path.exists(dialogues_filepath):
+            raise ValueError(f"Dialogues file {dialogues_filepath} does not exist")
+
+        logging.info(f"Reading dialogue examples from {dialogues_filepath}")
+        with open(dialogues_filepath, "r") as f:
+            data = json.load(f)["data"]
+
+        logging.info(f"Processing dialogues...")
+        examples = []
+        for dialogue_idx, dialogue in enumerate(tqdm(data)):
+
+            # skip dialogues where system slots are absent
+            if not dialogue["Fulfillment"]:
+                continue
+
+            DialogueCustomDataProcessor.add_metadata_to_fulfillment_slots(dialogue)
+            dialogue_dict = DialogueCustomDataProcessor.create_dialogue_sgd_format_dict(dialogue, dialogue_idx)
+            if dialogue_dict:
+                curr_examples = DialogueCustomDataProcessor.create_examples_from_dialogue_dict(dialogue_dict)
+                examples.extend(curr_examples)
+
+        return examples
+
+    @staticmethod
+    def get_user_entity_to_slot_mapping(dialogue):
+        """
+        Formats entities detected in user utterance into slots structure to match SGD format
+        """
+
+        entity_to_slot_mapping = {}
+        for slot in dialogue["Slots"]:
+            slot_name = slot["Name"]
+            if isinstance(slot["Values"][0], str):
+                slot_value = slot["Values"][0]
+            elif isinstance(slot["Values"][0], dict):
+                slot_value = slot["Values"][0]["Text"]
+
+            for entity in dialogue["Entities"]:
+                if entity["Token"] == slot_value:
+                    entity_to_slot_mapping.update({slot_name: entity})
+                    entity_to_slot_mapping[slot_name].update({"value": slot_value})
+                    break
+
+        return entity_to_slot_mapping
+
+    @staticmethod
+    def add_metadata_to_fulfillment_slots(dialogue):
+        """
+        Appends metadata from dialog state to current slots in the dialogue
+        """
+
+        for slot in dialogue["Fulfillment"][0]["FulfillmentSlots"]:
```

Comment:
Because of lack of an external schema, I don't think we can find out all slot names without going through the dialogues. I can create a map, and populate it with a slot name --> modified name whenever slot name is encountered first, and then just keep using this value to avoid the loop every time

Reply:
Sounds good!